FlintCore/Utils/CompletionRequirement.swift

```diff
@@ -0,0 +1,210 @@
+///
+/// Then make your function that requires a completion handler take an instance of this type instead of the closure type, and make
+/// the function expect a return value of the nested `Status` type:
+///
+/// ```
+/// protocol MyCoordinator {
+///   typealias DoSomethingCompletion = CompletionRequirement<Bool>
+///
+///   func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status
+/// }
+/// ```
+///
+/// Now, when calling this function on the protocol, you construct the requirement instance, pass it and verify the result:
+///
+/// ```
+/// let coordinator: MyCoordinator = ...
+/// let completion = MyCoordinator.DoSomethingCompletion( { (shouldCancel: Bool, completedAsync: Bool) in
+///    print("Cancel? \(shouldCancel)")
+/// })
+///
+/// The block takes one argument of type `T`, in this case a boolean, and a second `Bool` argument that indicates
+/// if the completion block has been called asynchronously.
+///
+/// // Call the function that requires completion
+/// let status = coordinator.doSomething(input: x, completionRequirement: completion)
+///
+/// // Make sure one of the valid statuses was returned.
+/// // This safety test ensures that the completion from the correct completion requirement instance was returned.
+/// precondition(completion.verify(status))
+///
+/// // If the result does not return true for `isCompletingAsync`, the completion callback will have already been called by now.
+/// if !status.isCompletingAsync {
+///     print("Completed synchronously: \(status.value)")
+/// } else {
+///     print("Completing asynchronously... see you later")
+/// }
+/// ```
+///
+/// When implementing such a function requiring a completion handler, you return one of two statuses returned by either
+/// the `CompletionRequirement.completed(_ arg: T)` or `CompletionRequirement.willCompleteAsync()`.
+/// The `CompletionRequirement` will take care of calling the completion block as appropriate.
+///
+/// ```
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     return completionRequirement.completedSync(false)
+/// }
+///
+/// // or for async completion
+///
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     let result = completionRequirement.willCompleteAsync()
+///     DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
+///         result.completed(false)
+///     }
+///     return result
+/// }
+/// ```
+public class CompletionRequirement<T> {
+    public class Status {
+        fileprivate let _value: T?
+        fileprivate var value: T {
+            get {
+                guard let result = _value else {
+                    flintBug("CompletionRequirement status value is nil")
+                }
+                return result
+            }
+        }
+        
+        init() {
+            _value = nil
+        }
```

Comment:
Wouldn't this always result in a `flintBug` from the above getter? Shouldn't this blank initializer just not exist? Or is this the way you show that there is a bug, so it has to exist? Seems like maybe this bug aspect should be reported before here instead of like this, no?

Reply:
Looking below maybe it's required because you have to call super? I feel like there may be a better way to achieve this? Maybe using a protocol for `Status` instead of doing inheritance? I haven't thought about it too much yet to have a specific idea. Let me know if you'd like to discuss further, or just continue with this path for now.